File: Interviews/GFG/Daily-Problems/2022/October/Day-08/Substrings-of-length-k-with-k-1-distinct-elements.cpp
class Solution {
public:
    int countOfSubstrings(string s, int k) {
        int n = s.size(), ans = 0, extra = 0;
        vector<int> cnt(26, 0);

        for (int i = 0; i < n; i++) {
            if (cnt[s[i] - 'a'] > 0) extra += 1;
            cnt[s[i] - 'a'] += 1;

            if (i >= k - 1) {
                if (extra == 1) ans += 1;

                if (cnt[s[i - k + 1] - 'a'] > 1) extra -= 1;
                cnt[s[i - k + 1] - 'a'] -= 1;
            }
        }

        return ans;
    }
};

File: Interviews/GFG/random-problems/Medium/Substrings-of-length-k-with-k-1-distinct-elements/Substrings-of-length-k-with-k-1-distinct-elements.py
class Solution:
    def countOfSubstrings(self, s, k):
        n, ans , extra, cnt = len(s), 0, 0, [0] * 26
        
        for i in range(n):
            if cnt[ord(s[i]) - 97] > 0: extra += 1
            cnt[ord(s[i]) - 97] += 1
            
            if i >= k - 1:
                if extra == 1: ans += 1
                if (cnt[ord(s[i - k + 1]) - 97] > 1): extra -= 1
                cnt[ord(s[i - k + 1]) - 97] -= 1
        
        return ans